src/servers/src/auth/user_provider.rs

```diff
@@ -0,0 +1,240 @@
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        let (mode, content) = value.split_once(':').context(InvalidConfigSnafu {
+            value: value.to_string(),
+            msg: "MemUserProviderOption must be in format `<option>:<value>`",
+        })?;
+        return match mode {
+            "file" => {
+                // check valid path
+                let path = Path::new(content);
+                if !path.exists() || !path.is_file() {
+                    return InvalidConfigSnafu {
+                        value: content.to_string(),
+                        msg: "MemUserProviderOption file path must be a valid file",
+                    }
+                    .fail();
+                }
+
+                let file = File::open(path).context(IOErrSnafu)?;
+                let credential = io::BufReader::new(file)
+                    .lines()
+                    .filter_map(|line| line.ok())
+                    .filter_map(|line| {
+                        if let Some((k, v)) = line.split_once('=') {
+                            Some((k.to_string(), v.as_bytes().to_vec()))
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<HashMap<String, Vec<u8>>>();
+                if credential.is_empty() {
+                    return InvalidConfigSnafu {
+                        value: content.to_string(),
+                        msg: "MemUserProviderOption file must contain at least one valid credential",
+                    }
+                    .fail();
+                }
+
+                Ok(MemUserProvider { users: credential, })
+            }
+            "inline" => content
+                .split(',')
+                .map(|kv| {
+                    let (k, v) = kv.split_once('=').context(InvalidConfigSnafu {
+                        value: kv.to_string(),
+                        msg: "MemUserProviderOption inline values must be in format `user=pwd[,user=pwd]`",
+                    })?;
+                    Ok((k.to_string(), v.as_bytes().to_vec()))
+                })
+                .collect::<Result<HashMap<String, Vec<u8>>, Error>>()
+                .map(|users| MemUserProvider { users }),
+            _ => InvalidConfigSnafu {
+                value: mode.to_string(),
+                msg: "MemUserProviderOption must be in format `file:<path>` or `inline:<values>`",
+            }
+            .fail(),
+        };
+    }
+}
+
+pub struct MemUserProvider {
+    users: HashMap<String, Vec<u8>>,
+}
+
+#[async_trait]
+impl UserProvider for MemUserProvider {
+    fn name(&self) -> &str {
+        "mem_user_provider"
+    }
+
+    async fn auth(
+        &self,
+        input_id: Identity<'_>,
+        input_pwd: Password<'_>,
+    ) -> Result<UserInfo, Error> {
+        match input_id {
+            Identity::UserId(username, _) => {
+                if let Some(save_pwd) = self.users.get(username) {
```

Comment:
```suggestion
                let save_pwd = self.users.get(username).context(UserNotFoundSnafu)?;
```
to save a layer of indention